File: Heap Priority Queue/0857. Minimum Cost to Hire K Workers/0857.cpp
#pragma GCC optimize("O3","Oz","unroll-loops","inline-functions","omit-frame-pointer","march=native","lto","fast-math")
class Solution {
public:
    double mincostToHireWorkers(vector<int>& quality, vector<int>& wage, int k) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        double res=DBL_MAX;
        int totalQuality=0;
        priority_queue<int>maxQualityHeap;

        vector<pair<double,int>>wageQuality;
        for(int i=0; i<wage.size(); ++i){
            wageQuality.push_back({double(wage[i])/quality[i],quality[i]});
        } sort(wageQuality.begin(),wageQuality.end());

        for(int i=0; i<wageQuality.size(); ++i){
            if(maxQualityHeap.size()==k){
                totalQuality-=maxQualityHeap.top();
                maxQualityHeap.pop();
            }
            totalQuality+=wageQuality[i].second;
            maxQualityHeap.push(wageQuality[i].second);
            if(maxQualityHeap.size()==k){
                res=min(res,totalQuality*wageQuality[i].first);
            }
        }
        return res;
    }
};

File: Heap Priority Queue/0857. Minimum Cost to Hire K Workers/0857.py
class Solution:
    def mincostToHireWorkers(self, quality: List[int], wage: List[int], k: int) -> float:
        wageQuality=[]
        for i in range(len(wage)):
            wageQuality.append((wage[i]/quality[i],quality[i]))
        wageQuality.sort()
        res=float('inf')
        currQuality=0

        maxHeap=[]

        for i in range(len(wageQuality)):
            if len(maxHeap)==k:
                currQuality-=abs(maxHeap[0])
                heapq.heappop(maxHeap)
            
            currQuality+=wageQuality[i][1]
            heapq.heappush(maxHeap,(-wageQuality[i][1]))
            if len(maxHeap)==k:
                res=min(res,wageQuality[i][0]*currQuality)

        return res
